larrecodnn/ImagePatternAlgs/Modules/RawWaveformClnSigDump_module.cc

```diff
@@ -0,0 +1,807 @@
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+// LArSoft libraries
+#include "larcore/Geometry/Geometry.h"
+#include "larcorealg/Geometry/PlaneGeo.h"
+#include "larcoreobj/SimpleTypesAndConstants/RawTypes.h" // raw::ChannelID_t
+#include "lardata/DetectorInfoServices/DetectorClocksService.h"
+#include "lardata/DetectorInfoServices/DetectorPropertiesService.h"
+#include "lardataobj/RawData/RawDigit.h"
+#include "lardataobj/RawData/raw.h"
+#include "lardataobj/RecoBase/Hit.h"
+#include "lardataobj/RecoBase/Wire.h"
+#include "lardataobj/Simulation/SimChannel.h"
+#include "larsim/MCCheater/ParticleInventoryService.h"
+#include "nusimdata/SimulationBase/MCParticle.h"
+#include "nusimdata/SimulationBase/MCTruth.h"
+
+#include "larevt/CalibrationDBI/Interface/ChannelStatusProvider.h"
+#include "larevt/CalibrationDBI/Interface/ChannelStatusService.h"
+
+#include "TRandom3.h"
+#include "c2numpy.h"
+
+using std::cout;
+using std::endl;
+using std::ofstream;
+using std::string;
+
+struct WireSigInfo {
+  int pdgcode;
+  std::string genlab;
+  std::string procid;
+  unsigned int tdcmin;
+  unsigned int tdcmax;
+  int tdcpeak;
+  int adcpeak;
+  int numel;
+  double edep;
+};
+
+namespace nnet {
+  class RawWaveformClnSigDump;
+}
+
+class nnet::RawWaveformClnSigDump : public art::EDAnalyzer {
+
+public:
+  explicit RawWaveformClnSigDump(fhicl::ParameterSet const& p);
+
+  // Plugins should not be copied or assigned.
+  RawWaveformClnSigDump(RawWaveformClnSigDump const&) = delete;
+  RawWaveformClnSigDump(RawWaveformClnSigDump&&) = delete;
+  RawWaveformClnSigDump& operator=(RawWaveformClnSigDump const&) = delete;
+  RawWaveformClnSigDump& operator=(RawWaveformClnSigDump&&) = delete;
+
+  // Required functions.
+  void analyze(art::Event const& e) override;
+
+  //void reconfigure(fhicl::ParameterSet const & p);
+
+  void beginJob() override;
+  void endJob() override;
+
+private:
+  std::string fDumpWaveformsFileName;
+  std::string fDumpCleanSignalFileName;
+
+  std::string fSimulationProducerLabel; ///< producer that tracked simulated part. through detector
+  std::string fSimChannelLabel;         ///< module that made simchannels
+  std::string fDigitModuleLabel;        ///< module that made digits
+  std::string fCleanSignalDigitModuleLabel; ///< module that made the signal-only digits
+  bool fUseFullWaveform;
+  unsigned int fShortWaveformSize;
+  int fEstIndFWForOffset;
+  int fEstColFWForOffset;
+
+  std::string fSelectGenLabel;
+  std::string fSelectProcID;
+  int fSelectPDGCode;
+  std::string fPlaneToDump;
+  double fMinParticleEnergyGeV;
+  double fMinEnergyDepositedMeV;
+  int fMinNumberOfElectrons;
+  int fMaxNumberOfElectrons;
+  int fMinPureSignalADCs;
+  bool fSaveSignal;
+  int fMaxNoiseChannelsPerEvent;
+  std::string fCollectionPlaneLabel;
+  art::ServiceHandle<geo::Geometry> fgeom;
+  art::ServiceHandle<cheat::ParticleInventoryService> PIS;
+
+  TRandom3 *fRand;
```

Comment:
is using the art random services not appropriate here?

Reply:
thanks for reviewing the code Dom, I replaced this now with CLHEP random and nurandom services.